ui/kotlinx-coroutines-android/test/HandlerDispatcherTest.kt

```diff
@@ -132,4 +132,11 @@ class HandlerDispatcherTest : TestBase() {
         mainLooper.scheduler.advanceBy(51, TimeUnit.MILLISECONDS)
         finish(5)
     }
+
+    @Test
+    fun testHandlerDispatcherNotEqualToImmediate() {
```

Comment:
This change changes the behaviour of the following code:
```
@Test
    fun foo() = runBlocking<Unit> {
        withContext(Dispatchers.Main) {
            println("1")
            withContext(Dispatchers.Main.immediate) {
                println("2")
                launch(Dispatchers.Main) {
                    println("4")
                }
                withContext(Dispatchers.Main) {
                    println("3")
                }
            }
        }
    }
```

It seems like this is expected (at least from the issue description), and I would rather expect the current (after your change) behaviour, but still pointing it out just in case

Reply:
Oh. Didn't realize this. Then, we'll have to weigh the pros and cons.

I checked all the other `Dispatchers.Main.immediate` implementations, and they all use object identity for equality checks, so this change is also needed simply for consistency.

Also, before this change, the behavior of
```kotlin
            withContext(Dispatchers.Main.immediate) {
                println("2")
                launch(Dispatchers.Main) {
                    println("4")
                }
                withContext(Dispatchers.Main) {
                    println("3")
                }
            }
```
was `2, 4, 3` if it was not called from the main thread, but `2, 3, 4` it it was. Now, it's consistently `2, 4, 3`. I think it's more intuitive that the calling thread doesn't affect the ordering.

Also, yes, I, too, would expect the proposed behavior:

```kotlin
fun updateUi() = scope.launch(Dispatchers.Main) { }

suspend fun readUiElement(): Int = withContext(Dispatchers.Main) { 1 }

fun update() = viewModelScope.launch {
  updateUi()
  cache = readUiElement()
}
```

I think it's reasonable to expect that `readUiElement` detects the results of `updateUi`. Currently, for this to work reliably, I think a `yield` would need to be inserted between the two lines.

... That said, of course, this will subtly break someone's code that relied on the existing implicit ordering.

I'd say the pros outweigh the cons. Most people would be better served by having the execution order roughly the same as the order of function invocations.

Reply:
Then it's worth doing the following:

* Fix this behaviour with tests explicitly (in all integrations)
* Maybe document it around `immediate` property
* Rephrase the issue (or create a new one) in a bit more clear manner so we'll reference it in the changelog properly

It's okay to do it both in and out of the scope of this issue. WDYT?